packages/react-components/src/gallery/dominantSpeaker.ts

```diff
@@ -0,0 +1,61 @@
+import { VideoGalleryRemoteParticipant } from '../types';
+
+/**
+ * Calculates the participants that should be renderd basis on the list of dominant speakers and currently renderd participants in a call.
+ * @param participants - Array containing all participants of a call. {@link @azure/communication-react#VideoGalleryRemoteParticipant}
+ * @param dominantSpeakers - An array containing the userId of dominant speakers in a call in the order of their dominance. 0th index is the most dominant, 1st is the second most etc
+ * @param visibleParticipants - Array containing currently rendered (visible) participants in the call. {@link @azure/communication-react#VideoGalleryRemoteParticipant}
+ * @param MAX_TILES - Maximum number of tiles to calculate.
+ * @returns VideoGalleryRemoteParticipant[] {@link @azure/communication-react#VideoGalleryRemoteParticipant}
+ */
+export const smartDominantSpeakerParticipants = (
+  participants: VideoGalleryRemoteParticipant[] = [],
+  dominantSpeakers: Array<string> = [],
+  visibleParticipants: VideoGalleryRemoteParticipant[] = [],
+  MAX_TILES = 4
+): VideoGalleryRemoteParticipant[] => {
+  if (!participants) return [];
+
+  // Only use the Max allowed dominant speakers.
+  const dominantSpeakerIds = dominantSpeakers.slice(0, MAX_TILES);
+
+  // Don't apply any logic if total number of video streams is less than Max video streams.
+  if (participants.length <= MAX_TILES) return participants;
+
+  // Initialize `visibleParticipants` if it is empty.
+  if (!visibleParticipants.length) {
+    visibleParticipants = participants.slice(0, MAX_TILES);
+  }
+
+  const lastVisibleSpeakerIds = visibleParticipants.map((speaker) => speaker.userId);
+  const newDominantSpeakerIds = dominantSpeakerIds.filter((id) => !lastVisibleSpeakerIds.includes(id));
+
+  // Remove participants that are no longer dominant and replace them with new dominant speakers.
+  lastVisibleSpeakerIds.forEach((id, idx) => {
+    if (!dominantSpeakerIds.includes(id)) {
+      const replacement = newDominantSpeakerIds.shift();
+      if (replacement) {
+        lastVisibleSpeakerIds[idx] = replacement;
+      }
+    }
+  });
+
+  // Sort the new video participants to match the order of last visible participants.
+  const videoParticipantsToRender = participants.filter((p) => lastVisibleSpeakerIds.includes(p.userId));
+  videoParticipantsToRender.sort((a, b) => {
+    return lastVisibleSpeakerIds.indexOf(a.userId) - lastVisibleSpeakerIds.indexOf(b.userId);
+  });
+
+  // Add additional participants to the final list of visible participants if the list has less than Max visible participants.
+  if (videoParticipantsToRender.length < MAX_TILES) {
+    const diff = MAX_TILES - videoParticipantsToRender.length;
+    for (let index = 0; index < diff; index++) {
+      const filler = participants.find((p) => !videoParticipantsToRender.includes(p));
+      filler && videoParticipantsToRender.push(filler);
+    }
+  }
+
+  return videoParticipantsToRender;
```

Comment:
I'm a little concerned about rendering performance.

If there are changes to dominant speaker list of participant list but in a way that none of the visible participants change, we still rerender the gallery. Is that a concern?

In the end, should we compare the final result with `previouslyVisibleParticipants` (assuming you've updated that to cache the whole return value instead of just the IDs), and if all the contents are the same, return `previouslyVisibleParticipants` instead of the newly constructed list with the same items?

Reply:
The constant re-renders are un-avoidable now since we are also using the isSpeaking feature. Which causes dominant speakers to change and hence re-render whenever dominant speakers change.